### PR TITLE
Custom difficulty parameters

### PR DESCRIPTION
**Pull request by FozzeY**
Implements #790. Additional command line parameters for customizing difficulty as per https://doomwiki.org/wiki/Skill_level.

**Doom**
- `-doubleammo` double ammo pickup rate. Demo incompatible.

**Heretic**
- `-fast` fast monsters. Demo incompatible.
- `-moreammo` 50% more ammo pickup rate. Demo incompatible.
- `-autohealth` automatic use of Quartz flasks and Mystic urns.
- `-keysloc` keys are displayed on the automap.

**Hexen**
- `-fast` fast monsters. Demo incompatible.
- `-moremana` 50% more mana pickup rate. Demo incompatible.
- `-autohealth` automatic use of Quartz flasks and Mystic urns.

**Strife**
- `-doubleammo` double ammo pickup rate. Demo incompatible.

### CODE REVIEW

File: src/heretic/g_game.c [hunk 1]
```diff
@@ -780,6 +780,27 @@ void G_DoLoadLevel(void)
     // [crispy] update the "singleplayer" variable
     CheckCrispySingleplayer(!demorecording && !demoplayback && !netgame);
 
+    // [crispy] more ammo
```
Comment (fabiangreffrath) [inline src/heretic/g_game.c:783]: Is the `-autohealth` parameter considered demo-safe in Heretic and Hexen?

Comment (fabiangreffrath) [reply src/heretic/g_game.c:783]: @FozzeY Hello?

Comment (FozzeY) [reply src/heretic/g_game.c:783]: Sorry, holidays! Turns out this parameter is not demo-safe - I have added a conditional for it, however, it is currently does not prevent demo recording due to #807.

File: src/crispy.h [hunk 2]
```diff
@@ -87,13 +87,17 @@ typedef struct
 	int demowarp;
 	int fps;
 
+    boolean autohealth;
```
Comment (fabiangreffrath) [inline src/crispy.h:90]: Please keep indentation consistent with the surrounding lines.

File: src/doom/p_inter.c [hunk 3]
```diff
@@ -86,10 +86,12 @@ P_GiveAmmo
 	num = clipammo[ammo]/2;
     
     if (gameskill == sk_baby
-	|| gameskill == sk_nightmare)
+	|| gameskill == sk_nightmare
+    || (crispy->moreammo && !demoplayback))
     {
 	// give double ammo in trainer mode,
 	// you'll need in nightmare
+    // [crispy] double ammo
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c:94]: This comment is moot, the code was there before this change.

File: src/doom/p_inter.c [hunk 4]
```diff
@@ -86,10 +86,12 @@ P_GiveAmmo
 	num = clipammo[ammo]/2;
     
     if (gameskill == sk_baby
-	|| gameskill == sk_nightmare)
+	|| gameskill == sk_nightmare
+    || (crispy->moreammo && !demoplayback))
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c:90]: `critical->moreammo` please without the extra check.

File: src/heretic/g_game.c (hunk 1 above)
Comment (fabiangreffrath) [reply src/heretic/g_game.c:783]: Seems we cannot go with a single `crispy->singleplayer` variable in Heretic and Hexen. Please remove the `CheckCrispySingleplayer()` line above and change the check in the `crispy->pistolstart` block from `crispy->singleplayer` to `!demorecording && !demoplayback && !netgame`.
Please move the additional checks for the extra parameters to the beginning of `D_DoomLoop()` and change the `!crispy->singleplayer` checks to `!demorecording && gameaction != ga_playdemo && !netgame`. This applies to both Heretic and Hexen.

File: src/heretic/g_game.c [hunk 5]
```diff
@@ -1861,7 +1892,8 @@ void G_InitNew(skill_t skill, int episode, int map)
         respawnmonsters = false;
     }
     // Set monster missile speeds
-    speed = skill == sk_nightmare;
+    // [crispy] fast monsters
```
Comment (fabiangreffrath) [inline src/heretic/g_game.c:1895]: moot comment

File: src/heretic/p_enemy.c [hunk 6]
```diff
@@ -697,8 +697,9 @@ void A_Chase(mobj_t * actor)
         actor->threshold--;
     }
 
-    if (gameskill == sk_nightmare)
+    if (gameskill == sk_nightmare || (crispy->fast && !demoplayback))
     {                           // Monsters move faster in nightmare mode
+                                // [crispy] fast monsters
```
Comment (fabiangreffrath) [inline src/heretic/p_enemy.c:702]: moot comment

File: src/heretic/p_enemy.c [hunk 7]
```diff
@@ -735,11 +736,12 @@ void A_Chase(mobj_t * actor)
 
 //
 // don't attack twice in a row
-//
+// [crispy] fast monsters
```
Comment (fabiangreffrath) [inline src/heretic/p_enemy.c:739]: moot comment

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Sorry, but this requires some rework. Also, please remove all comments to code for which you merely add another condition.

File: src/doom/p_inter.c (hunk 4 above)
Comment (FozzeY) [reply src/doom/p_inter.c:90]: Is that not going to desync the built-in demos?

Comment (fabiangreffrath) [reply src/doom/p_inter.c:90]: The `critical` struct takes care of that. It is the same as the `crispy` but all zero durin demo and network sessions. But you are probably right, it is a lot easier to understand if you leave it as it is.

File: src/heretic/g_game.c (hunk 1 above)
Comment (FozzeY) [reply src/heretic/g_game.c:783]: The issue is that at this point `demorecording` is still `false`, so we cannot rely on it either. Perhaps setting the `demorecording = true;` before the `G_InitNew` call in `G_RecordDemo` would be feasible?
https://github.com/fabiangreffrath/crispy-doom/blob/05b426763b6381a09178d4ac1e986f99d176452c/src/heretic/g_game.c#L2094-L2151

Comment (fabiangreffrath) [reply src/heretic/g_game.c:783]: That's why I asked you to "please move the additional checks for the extra parameters to the beginning of D_DoomLoop() and change the !crispy->singleplayer checks to !demorecording && gameaction != ga_playdemo && !netgame." I know it's confusing! :wink:

**Comment by FozzeY** [comment]
Since the issue was `demorecording` and not `CheckCrispySingleplayer()` itself, I think it can still be used in the beginning of `D_DoomLoop()`. Seems to work fine on tests.

File: src/heretic/p_enemy.c [hunk 8]
```diff
@@ -735,11 +735,11 @@ void A_Chase(mobj_t * actor)
 
 //
 // don't attack twice in a row
-//
```
Comment (fabiangreffrath) [inline src/heretic/p_enemy.c:738]: Please keep this line.

File: src/heretic/p_enemy.c [hunk 9]
```diff
@@ -757,10 +757,10 @@ void A_Chase(mobj_t * actor)
 
 //
 // check for missile attack
-//
```
Comment (fabiangreffrath) [inline src/heretic/p_enemy.c:760]: Please keep this line.

File: src/hexen/p_enemy.c [hunk 10]
```diff
@@ -2019,7 +2021,7 @@ void A_SerpentChase(mobj_t * actor)
         actor->threshold--;
     }
 
-    if (gameskill == sk_nightmare)
+    if (gameskill == sk_nightmare || (critical->fast))
```
Comment (fabiangreffrath) [inline src/hexen/p_enemy.c:2024]: Why the extra parentheses here and in the following checks?

**Review by fabiangreffrath — COMMENTED** [review]
A few minor cosmetic issues left, but also one general one: The `crispy` and `critical` structs are not updated automatically, they are only updated when `CheckCrispySingleplayer()` is called. So, for the non-Doom games it is alright to call this at the beginning of `D_DoomLoop()` to catch the demo recording, single demo playback and network game cases. But we currently have no means to detect playback of demos from the in-game loop. One possible approach could be to say "well, if you pass these parameters, don't expect internal demos to keep sync", but I'd rather prefer a cleaner solution. :wink: Any idea?

**Comment by fabiangreffrath** [comment]
@FozzeY Have you given up on this? I hope not...

**Comment by FozzeY** [comment]
@fabiangreffrath Not at all! Apologies it is taking a while, I want to make sure my solution is optimal and thoroughly tested too.

**Comment by fabiangreffrath** [comment]
Thank you, no need to apologize! I am fine if the games (all 4) refuse to start if one of the new command line parameters is given together with one of `-record`, `-playdemo`, `-fastdemo`, `-connect`, `-solo-net`, etc. - you get the point. I think this is how it currently even works. In-game demos from the demo reel are seocnd priority.

I am going to restructure some of the code anyway... :wink:

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
One last request:
If you introduce a new check into an existing sequence of checks, could you please add that new check last in the sequence? For example, let this
```
 if (gameskill < sk_nightmare && actor->movecount)
```
become this
```
if (gameskill < sk_nightmare && actor->movecount && !critical->fast )
```
instead of this
```
if (gameskill < sk_nightmare &&
    !critical->fast && actor->movecount)
```
That's all. 😉

**Review by fabiangreffrath — APPROVED** [review]

**Comment by fabiangreffrath** [comment]
21 commits! Phew, thank you for your patience!

**Pull request merged**